go_modules/lib/dependabot/go_modules/version.rb

```diff
@@ -19,41 +21,48 @@ class Version < Dependabot::Version
                         '(\+incompatible)?'
       ANCHORED_VERSION_PATTERN = /\A\s*(#{VERSION_PATTERN})?\s*\z/
 
+      sig { override.params(version: VersionParameter).returns(T::Boolean) }
       def self.correct?(version)
         version = version.gsub(/^v/, "") if version.is_a?(String)
         version = version.to_s.split("+").first if version.to_s.include?("+")
 
         super(version)
       end
 
+      sig { override.params(version: VersionParameter).void }
       def initialize(version)
-        @version_string = version.to_s.gsub(/^v/, "")
+        @version_string = T.let(version.to_s.gsub(/^v/, ""), String)
         version = version.gsub(/^v/, "") if version.is_a?(String)
         version = version.to_s.split("+").first if version.to_s.include?("+")
+        @prerelease = T.let(nil, T.nilable(String))
         version, @prerelease = version.to_s.split("-", 2) if version.to_s.include?("-")
 
         super
       end
 
+      sig { returns(String) }
       def inspect # :nodoc:
         "#<#{self.class} #{@version_string.inspect}>"
       end
 
+      sig { returns(String) }
       def to_s
         @version_string
       end
 
+      sig { params(other: Object).returns(T.nilable(Integer)) }
       def <=>(other)
         result = super(other)
         return if result.nil?
         return result unless result.zero?
 
-        other = self.class.new(other) unless other.is_a?(Version)
+        other = self.class.new(other.to_s) unless other.is_a?(Version)
```

Comment:
Was this a bug discovered through the typing exercise?

Reply:
So the `<=>` method comes from `kernel.rbi` and is defined as

```ruby
sig do
    params(
        other: BasicObject,
    )
    .returns(T.nilable(Integer))
  end
  def <=>(other); end
```

Unfortunately, overriding that directly results in 2 errors:

> Expected `T.nilable(T.any(String, Integer, Gem::Version))` but found `BasicObject` for argument `version`[7002](https://srb.help/7002)

and 

> Method `is_a?` does not exist on `BasicObject` (fix available)[7003](https://srb.help/7003)

I think I could force it to use `VersionParameter` to fix both these, but I am not 100% sure if I can do that. This felt like a compromise.